File: playgroundPython/recursiveStringReversal.py
"""
Recursive String Reversal:
Write a recursive function to reverse a given string without using additional data structures.

Note: Python strings are immutable, so true in-place swapping (like the C
version) isn't possible here. This mirrors the C++ approach instead: peel
off the first character, recurse on the rest, then append it at the end.
"""


def reverse_string(s: str) -> str:
    if len(s) <= 1:
        return s
    first_char = s[0]
    rest = s[1:]
    return reverse_string(rest) + first_char


if __name__ == "__main__":
    word = "hej"
    print(f"Original: {word}")
    print(f"Reversed: {reverse_string(word)}")

File: playgroundC/recursiveStringReversal.c
#include <stdio.h>
#include <string.h>

// Recursive String Reversal:
// Write a recursive function to reverse a given string without using additional data structures.
//
// Unlike the C++ version (which builds a new std::string on each call via
// substr), this reverses the char array in place: swap the first and last
// character, then recurse inward on the range between them. No extra
// buffer/array is allocated - just index bookkeeping and one temp char.

void reverseHelper(char *str, int start, int end) {
    if (start >= end) {
        return; // base case: pointers have met or crossed
    }

    char temp = str[start];
    str[start] = str[end];
    str[end] = temp;

    reverseHelper(str, start + 1, end - 1);
}

void reverseString(char *str) {
    reverseHelper(str, 0, (int)strlen(str) - 1);
}

// RECURSIVE_TESTING is defined by tests/test_recursiveStringReversal.c
// before it #includes this file, so the test binary (which has its own
// main()) doesn't end up with two main() functions in one link.
#ifndef RECURSIVE_TESTING
int main(void) {
    char word[] = "hej";

    printf("Original: %s\n", word);
    reverseString(word);
    printf("Reversed: %s\n", word);

    return 0;
}
#endif
